core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -15,160 +15,184 @@
-                retryBackoff.nextDelayMillis(1), TimeUnit.MILLISECONDS);
-    }
+        // Stop the health checkers in parallel.
+        final CompletableFuture<List<Object>> stopFutures;
+        synchronized (contexts) {
+            stopFutures = CompletableFutures.allAsList(
+                    contexts.values().stream()
+                            .map(ctx -> ctx.destroy().exceptionally(cause -> {
+                                logger.warn("Failed to stop a health checker for: {}",
+                                            ctx.endpoint(), cause);
+                                return null;
+                            }))
+                            .collect(toImmutableList()));
+            contexts.clear();
+        }
 
-    private CompletableFuture<Void> checkAndUpdateHealthyServers() {
-        final List<ServerConnection> checkedServers = updateServerList();
-
-        final CompletableFuture<List<Boolean>> healthCheckResults = CompletableFutures.successfulAsList(
-                checkedServers.stream()
-                              .map(connection -> connection.healthChecker.isHealthy(connection.endpoint()))
-                              .collect(toImmutableList()),
-                t -> false);
-        return healthCheckResults.handle((result, thrown) -> {
-            final ImmutableList.Builder<Endpoint> newHealthyEndpoints = ImmutableList.builder();
-            for (int i = 0; i < result.size(); i++) {
-                if (result.get(i)) {
-                    newHealthyEndpoints.add(checkedServers.get(i).endpoint());
-                }
-            }
-            setEndpoints(newHealthyEndpoints.build());
-            return null;
-        });
-    }
+        super.close();
+        delegate.close();
 
-    /**
-     * Update the servers this health checker client talks to.
-     */
-    private List<ServerConnection> updateServerList() {
-        final Map<Endpoint, ServerConnection> allServersByEndpoint = allServers
-                .stream()
-                .collect(toImmutableMap(ServerConnection::endpoint,
-                                        Function.identity(),
-                                        (l, r) -> l));
-        return allServers = delegate
-                .endpoints()
-                .stream()
-                .map(endpoint -> {
-                    ServerConnection connection = allServersByEndpoint.get(endpoint);
-                    if (connection != null) {
-                        return connection;
-                    }
-                    return new ServerConnection(endpoint, createEndpointHealthChecker(endpoint));
-                })
-                .collect(toImmutableList());
+        // Wait until the health checkers are fully stopped.
+        try {
+            stopFutures.join();
+        } catch (Exception e) {
+            logger.warn("Failed to stop all health checkers:", e);
+        }
```

Comment:
So we don't need this try catch anymore. 😉

Reply:
Ah, yeah. Nice catch.